### ISSUE TITLE
[0.6.0][Core] - Good bye internal option-indexing

### ISSUE THREAD

**Issue by SamBrishes** [pytesNET/tail.select#89]
Hellow,

I'm just working on version 1.0.0 of the **tail.select** script, where I currently try to remove the internal used option indexing, completely. This will change many methods and beahviours on the default `options` class, but I guess it is the best to handle the complete option structure using the source `<select>` field itself (possibly with minimal caching through the class itself.)

Since I'm translating the complete stuff in TypeScript, I'm just at the beginning, but the following method shows the new handler for `tail.select.options.get()`, which allows now a few more settings:

```typescript
/*
 |  OPTIONS :: GET ONE OR MORE ITEMS
 |  @since  1.0.0
 |
 |  @param  multi   The option value or the position number (as used in the source <select>)
 |                  Use null to don't restrict the selection.
 |  @param  multi   The <optgroup> label (Use '#' for ungrouped options).
 |                  Use null to don't restrict the selection.
 |  @param  bool    TRUE to exclude disabled options
 |  @param  bool    TRUE to exclude selected options
 |  @param  bool    TRUE to exclude hidden options
 |
 |  @return multi   A NodeList with all available <option> HTMLElements, or null if no 
 |                  valid option could be found or FALSE on failure.
 */
get(value: any, group: any, disabled: boolean = false, checked: boolean = false, hidden: boolean = true) {
    let selector = (hidden? ":not([hidden])": "") 
                 + (checked? ":not(:checked)": "") 
                 + (disabled? ":not(:disabled)": "");

    // Get all Items
    if(value === null && group === null) {
        var items: any = this.source.querySelectorAll(`option${selector}`);
    }

    // Get Items by Value or Number
    if(typeof value === "string" || typeof value === "number"){
        if(typeof value === "string"){
            selector = `option[value='${value}']${selector}`;
        } else {
            let nth = (value > 0)? "option:nth-child": "option:nth-last-child";
            selector = `${nth}(${Math.abs(value)})${selector}`;
        }

        if(typeof group !== "string"){
            var items: any = this.source.querySelectorAll(selector);
        } else if(typeof group === "string" && group !== "#"){
            var items: any = this.source.querySelectorAll(`optgroup[label='${group}'] ${selector}`);
        } else {
            var items: any = Array.prototype.filter.call(this.source.children, (item) => {
                return item.matches(selector);
            });
        }
    }

    // Get Items by Group
    if(value === null && typeof(group) === "string") {
        if(group === "#"){
            var items: any = Array.prototype.filter.call(this.source.children, (item) => {
                return item.matches(selector);
            });
        } else {
            var items: any = this.source.querySelectorAll(`optgroup[label='${group}'] option${selector}`);
        }
    }

    // Return
    if(typeof items === "undefined"){
        return false;
    }
    return (items.length > 0)? items: null;
}
```

Sincerely,
Sam.

PS: The code is untested :P

**Comment by Calak** [comment]
For selected/disabled/hidden you could instead accept true/false/null.
- true: only option that have the attribute setted (selected, disabled or hidden)
- false: only option where attribute isn't set on  (unselected, not disabled or not hidden)
- null: don't care about selected, disabled or hidden attributes.

**Comment by Calak** [comment]
Also, you could accept an object as parameter (referring to SearchConfig), allowing to omits parameters instead of explicitly passing null.

**Comment by Calak** [comment]
Did you checked support for :checked ? (sig)

```[selected=selected]``` could be an alternative (not tested), but if the support for ```:checked``` is enough for you, it's better and more scalable, because it allow to, later, falling back on checkboxes/radios to generate a list if we want to handle inputs as source. (If you see what I mean).

**Comment by Calak** [comment]
Did you checked support for :checked ? (sig)

```[selected=selected]``` could be an alternative (not tested), but if the support for ```:checked``` is enough for you, it's better and more scalable, because it allow to, later, falling back on checkboxes/radios to generate a list if we want to handle inputs as source. (If you see what I mean).

**Comment by SamBrishes** [comment]
Hellow,

The `true`, `false`, `null` ability is a good idea.

`:checked` has been shipped with the [(CSS 3) Selectors Level 3](https://caniuse.com/#search=%3Achecked), which has been taken over to JavaScript and is available [since IE9](https://technet.microsoft.com/en-us/jj127296(v=vs.71)). And it is still the ["W3-official"](https://www.w3.org/TR/html50/disabled-elements.html#selector-checked) way to go, so the support is widely given. :D

Sincerely,
Sam.

**Comment by SamBrishes** [comment]
Hellow,

still untested, but I updated this method with the following changes:

- A little more compact without loss of functionality
- Allows now passing `Element`, `Array`, `NodeList` and `HTMLCollection` as first argument, which gets filtered if another argument is present.
    - A single `Element` object always gets array'd to prevent bugs on other methods.
- A better / fixed workflow if you pass a number as first argument and a group or "#"

```typescript
/*
 |  OPTIONS :: GET ONE OR MORE ITEMS
 |  @since  1.0.0
 |
 |  @param  multi   The <option>s value as STRING, or the respective order number (use negative
 |                  numbers to count from the bottom) as listed in the source <select> field.
 |                  You can also pass null to don't restrict the selection to a specific value.
 |                  ---
 |                  A single <option> Element object, an Array with Elements, a NodeList or a
 |                  HTMLCollection gets filtered if you pass one of the following 4 arguments,
 |                  otherwise it gets returned as it is, just single Elements gets "array'd".
 |  @param  multi   The <optgroup> label STRING, use '#' for ungrouped <option>s only.
 |                  Use null to don't restrict the selection to an <optgroup>.
 |  @param  bool    TRUE to exclude disabled options, FALSE to explizit get them.
 |  @param  bool    TRUE to exclude selected options, FALSE to explizit get them.
 |  @param  bool    TRUE to exclude hidden options, FALSE to explizit get them.
 |
 |  @return multi   All available <option> Element objects as Array, NodeList or HTMLCollection.
 |                  The returning value can also contain a length of 0 if nothing could found.              
 */
get(value: any, group: any, disabled: boolean, checked: boolean, hidden: boolean) {
    let selector = (hidden)? "[hidden]": (hidden === false)? ":not([hidden])": ""
                 + (checked)? ":checked": (checked === false)? ":not(:checked)": ""
                 + (disabled)? ":disabled": (disabled === false)? ":not(:disabled)": "";

    // Filter Options
    if(value instanceof Element || "length" in value) {
        value = (value instanceof Element)? [Element]: value;
        return (selector === "" && group === void 0)? value: this.filter(value, group, selector);
    }

    // Get Options
    if(typeof value === "number") {
        let _nth = (value > 0)? ":nth-child": ":nth-last-child";
        selector = `option${_nth}(${Math.abs(value)})${selector}`;
    } else {
        selector = `option` + (value === void 0? ``: `[value="${value}"]`) + selector;
    }

    if(group === void 0){
        return this.source.querySelectorAll(selector);
    } else if(typeof group === "string" && group !== "#") {
        return this.source.querySelectorAll(`optgroup[label="${group}"] ${selector}`);
    } else if(group === "#" && typeof value !== "number") {
        return [].filter.call(this.source.children, (item) => item.matches(selector) );
    } else if(group === "#") {
        selector = `select[data-tail-select="${this.parent.id}"] > ${selector}`;
        return this.source.parentElement.querySelectorAll(selector);
    }

    // Invalid Parameters
    return [];
}
```

**Comment by Calak** [comment]
Are you sure about `value = (value instanceof Element)? [Element]: value;` ?

**Comment by SamBrishes** [comment]
Fixed ;D 
Thanks, didn't see that.

**Comment by Calak** [comment]
IMHO, use `value` for both filtering by exact values, filtering by positions and as a source container is a bad design and a bit confusing.
You could provide an additional `source` parameter that would be used in place of `this.source` if given. It would be more logical.

**Comment by SamBrishes** [comment]
Maybe the description is a bit confusing. `value` cannot be used to pass a (custom) source `<select>` container or to get / evaluate `custom` options, it still depends on `this.source`, which points to `tail.select.e` (so the main source `<select>` field).

The idea behind was to allow passing `<option>` HTMLElement objects of the same source `<select>` field (as of the tail.select instance) and return them as they are (unless the user specify another parameter, which is just an optional filtering feature).

That's really important, since the most methods of the `Options` class are using `.get()` to evaluate the respective item(s) (also in the current version!). And a few methods (such as `.remove()`) calling another method(s) of the `Options` class, which results in multiple calls of the `.get()` method (for already evaluated items). 

To decrease working time (by filtering already evaluated options) and increase the performance, I decided to "skip" `<option>` elements completely and returning them without additional checks. Of course, this allows tail.select users also to select desired `<option>` elements of the source `<select>` field by himself and just passing them to `tail.select.options.select`, for example.

And of course, there is the problem, that he can pass ANY `<option>` element of ANY `<select>` field... But we are talking about a developer feature, not something that can be exploited by the End-End-User.

**Comment by SamBrishes** [comment]
Done, will be published in 0.6.0
